File: python/Problem17.py
# 17. Letter Combinations of a Phone Number
# URL: https://leetcode.com/problems/letter-combinations-of-a-phone-numbe
# Difficulty: Medium
# Tags: Array, Recursion, Combinations

from typing import List

class Solution:
  digitToLetters: dict = {
    '2': "abc",
    '3': "def",
    '4': "ghi",
    '5': "jkl",
    '6': "mno",
    '7': "pqrs",
    '8': "tuv",
    '9': "wxyz"
  }

  def letterCombinations(self, digits: str) -> List[str]:
    results: List[str] = []
    if len(digits) > 0:
      self.combinations(results, digits, 0, "")
    return results
  
  def combinations(self, results: List[str], digits: str, index: int, s: str):
    if index >= len(digits):
      results.append(s)
      return
    
    letters: str = self.digitToLetters[digits[index]]
    for letter in letters:
      self.combinations(results, digits, index + 1, s + letter)

File: cpp/Problem17.cpp
/*
 * 17. Letter Combinations of a Phone Number
 * URL: https://leetcode.com/problems/letter-combinations-of-a-phone-numbe
 * Difficulty: Medium
 * Tags: Array, Recursion, Combinations
 */

#include <string>
#include <vector>
#include <map>

using namespace std;

class Solution {
public:
	vector<string> letterCombinations(string digits) {
		digitToLetters['2'] = "abc";
		digitToLetters['3'] = "def";
		digitToLetters['4'] = "ghi";
		digitToLetters['5'] = "jkl";
		digitToLetters['6'] = "mno";
		digitToLetters['7'] = "pqrs";
		digitToLetters['8'] = "tuv";
		digitToLetters['9'] = "wxyz";

		vector<string> results;
		if (digits.length() > 0) {
			string s = "";
			combinations(results, digits, 0, s);
		}
		return results;
	}

	void combinations(vector<string> &results, string &digits, int index, string &s) {
		if (index >= digits.length()) {
			results.push_back(s);
			return;
		}

		string letters = digitToLetters[digits.at(index)];
		for (int i = 0, length = letters.length(); i < length; i++) {
			s.push_back(letters.at(i));
			combinations(results, digits, index + 1, s);
			s.pop_back();
		}
	}
private:
	map<char, string> digitToLetters;
};